docs/contributing/doc_guide.rst

```diff
@@ -946,12 +946,67 @@ Docstring guidelines
   derivations and extensive discussions of mathematics in the "Notes"
   section.
 
-* Docstrings may include a Raises_ section that describes which
-  exceptions get raised and under what conditions. This section should
-  be used only for errors that are non-obvious or have a large chance of
-  getting raised. Documenting exceptions makes it easier for future
-  developers to plan exception handling. However, in practice this
-  section often becomes outdated.
+* Describe each :term:`parameter` in the "Parameters_" section of the
+  docstring.
+
+  .. code-block:: rst
+
+     Parameters
+     ----------
+     x : float
+        Description of ``x``.
+
+     y : bool, optional, default: True
+        Description of ``y``.
+
+  * Place type information to the right of the parameter name, along
+    with ``optional`` and/or ``keyword-only`` if necessary.
+
+  * Describe any requirements for the parameter, including preconditions
+    specified using |validate_quantities| or |particle_input|.
+
+  * Use the substitution ``|array_like|`` to indicate that an argument
+    should be able to be converted into an |ndarray|.
+
+  * If the shapes and sizes of the parameters are interrelated, then
+    include that information in parentheses immediately before the type
+    information. Include a trailing comma inside the parentheses when
+    the parameter is 1D.
+
+    .. code-block:: rst
+
+       Parameters
+       ----------
+       a : (M,) array_like
+           First input vector.
+
+       b : (N,) array_like
+           Second input vector.
+
+       out : (M, N) ndarray, optional
+           A location where the result is stored.
+
+    Use an asterisk (``*``) for a dimension that can be of arbitrary
+    size, and an ellipsis (``...``) to represent an arbitrary number of
+    dimensions that can each be of arbitrary size.
```

Comment:
This is adapted from https://github.com/numpy/numpydoc/pull/423, which includes a discussion of the motivation behind this syntax.  It's also done this way for a lot of the functions in NumPy (like [`numpy.outer`](https://numpy.org/doc/stable/reference/generated/numpy.outer)).  That PR doesn't include the `*` and `...` parts yet though.

Reply:
Also this is more consistent with the [matplotlib doc guide](https://matplotlib.org/stable/devel/documenting_mpl.html#parameter-type-descriptions).